onnxruntime/core/optimizer/layer_norm_fusion.cc

```diff
@@ -535,7 +535,12 @@ Status SimplifiedLayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int gr
 
     const Node* p_pow_input_node = graph_utils::GetInputNode(pow_node, 0);
     bool has_leading_cast = false;
-    if (p_pow_input_node) {
+    // If it's not GPU EP, since the CPU impl for SimplifiedLayerNormalization doesn't support input and scale
+    // having different types for now, and it may also have conflict to InsertCastTransformer,
+    // so the sub-graph will not be fused if it contains Cast Op.
+    bool is_gpu_ep = pow_node.GetExecutionProviderType() == kCudaExecutionProvider ||
```

Comment:
Can you explain this change a bit more? Based on the comment I would have expected the code to look for a Cast and exit if the CPU EP was involved.

I don't quite understand why we change the first branch which seems to be about setting `has_leading_cast` and a second location which has nothing to do with `has_leading_cast`.

Reply:
There are only 4 possible cases (x=Pow->ReduceMean->Add->Sqrt->Div, and y=Mul):
1) cast(to:float)->x->cast(to:fp16)->y : SimplifiedLayerNorm(T:fp16,V:fp16)
2) cast(to:float)->x->y : SimplifiedLayerNorm(T:fp16,V:float)
3) x->cast(to:fp16)->y : SimplifiedLayerNorm(T:float,V:fp16)
4) x->y : SimplifiedlayerNorm(T:float,V:float)

They all work for CUDA EP.

For CPU EP, we have only SimplifiedlayerNorm(T:float,V:float), so only #4 works. But if for #1 and #2, we just treat the entry cast as a normal node, means has_leading_cast=false, then for #2, we can still fuse it to "cast(to:float)->SimplifiedlayerNorm(T:float,V:float)" (just like applying #4 to the x->y after cast), so the condition for CPU EP to fuse or not is just to check if there is a cast between x and y.

Reply:
Would be great to put this excellent explanation in the comment so it's captured for the next person who works on the code.